akka-distributed-data/src/main/scala/akka/cluster/ddata/GCounter.scala

```diff
@@ -114,13 +114,12 @@ final class GCounter private[akka] (
     if ((this eq that) || that.isAncestorOf(this)) this.clearAncestor()
     else if (this.isAncestorOf(that)) that.clearAncestor()
     else {
-      var merged = that.state
-      for ((key, thisValue) <- state) {
-        val thatValue = merged.getOrElse(key, Zero)
-        if (thisValue > thatValue)
-          merged = merged.updated(key, thisValue)
-      }
       clearAncestor()
+      val merged = (state.keys ++ that.state.keys).map { key =>
+        val thisValue = state.getOrElse(key, Zero)
+        val thatValue = that.state.getOrElse(key, Zero)
+        (key, thisValue.max(thatValue))
+      }.toMap
       new GCounter(merged)
```

Comment:
Some things in Akka are optimized for performance and not for best idiomatic code style.
Seems like your suggested implementation is an order of magnitude slower than than before in a JMH bench that I wrote (and that is not even using counters with many entries).

BEFORE:
```
[info] Benchmark                                                                  Mode  Cnt      Score      Error   Units
[info] GCounterBenchmark.mergeAddFromBothNodes                                   thrpt    5  10067,158 ±  301,480  ops/ms
[info] GCounterBenchmark.mergeAddFromBothNodes:·gc.alloc.rate                    thrpt    5   1096,773 ±   32,797  MB/sec
[info] GCounterBenchmark.mergeAddFromBothNodes:·gc.alloc.rate.norm               thrpt    5    120,000 ±    0,001    B/op

[info] GCounterBenchmark.mergeAddFromOtherNode                                   thrpt    5   9048,687 ± 6984,785  ops/ms
[info] GCounterBenchmark.mergeAddFromOtherNode:·gc.alloc.rate                    thrpt    5    657,279 ±  507,201  MB/sec
[info] GCounterBenchmark.mergeAddFromOtherNode:·gc.alloc.rate.norm               thrpt    5     80,000 ±    0,001    B/op

[info] GCounterBenchmark.mergeAddFromSameNode                                    thrpt    5   5954,895 ±  203,640  ops/ms
[info] GCounterBenchmark.mergeAddFromSameNode:·gc.alloc.rate                     thrpt    5   5536,122 ±  187,876  MB/sec
[info] GCounterBenchmark.mergeAddFromSameNode:·gc.alloc.rate.norm                thrpt    5   1024,000 ±    0,001    B/op

[info] GCounterBenchmark.mergeComplex                                            thrpt    5   2226,622 ±   35,664  ops/ms
[info] GCounterBenchmark.mergeComplex:·gc.alloc.rate                             thrpt    5   1212,862 ±   19,265  MB/sec
[info] GCounterBenchmark.mergeComplex:·gc.alloc.rate.norm                        thrpt    5    600,000 ±    0,001    B/op
```

AFTER:
```
[info] Benchmark                                                                  Mode  Cnt     Score     Error   Units
[info] GCounterBenchmark.mergeAddFromBothNodes                                   thrpt    5  1111,977 ±  58,210  ops/ms
[info] GCounterBenchmark.mergeAddFromBothNodes:·gc.alloc.rate                    thrpt    5   928,774 ±  48,773  MB/sec
[info] GCounterBenchmark.mergeAddFromBothNodes:·gc.alloc.rate.norm               thrpt    5   920,000 ±   0,001    B/op

[info] GCounterBenchmark.mergeAddFromOtherNode                                   thrpt    5   939,763 ±  35,134  ops/ms
[info] GCounterBenchmark.mergeAddFromOtherNode:·gc.alloc.rate                    thrpt    5   771,417 ±  28,797  MB/sec
[info] GCounterBenchmark.mergeAddFromOtherNode:·gc.alloc.rate.norm               thrpt    5   904,000 ±   0,001    B/op

[info] GCounterBenchmark.mergeAddFromSameNode                                    thrpt    5  5917,432 ± 193,602  ops/ms
[info] GCounterBenchmark.mergeAddFromSameNode:·gc.alloc.rate                     thrpt    5  5501,810 ± 182,202  MB/sec
[info] GCounterBenchmark.mergeAddFromSameNode:·gc.alloc.rate.norm                thrpt    5  1024,000 ±   0,001    B/op

[info] GCounterBenchmark.mergeComplex                                            thrpt    5   421,577 ±  16,536  ops/ms
[info] GCounterBenchmark.mergeComplex:·gc.alloc.rate                             thrpt    5  1093,124 ±  43,063  MB/sec
[info] GCounterBenchmark.mergeComplex:·gc.alloc.rate.norm                        thrpt    5  2856,000 ±   0,001    B/op
```

Reply:
@patriknw Can you plz opening an PR for the Benchmark code? This way I can play with it :)

Reply:
@YikSanChan here is what I used https://github.com/akka/akka/blob/wip-GCounterBenchmark-patriknw/akka-bench-jmh/src/main/scala/akka/cluster/ddata/GCounterBenchmark.scala
Run with `jmh:run -prof gc -i 5 -wi 3 -f1 -t1 akka.cluster.ddata.GCounterBenchmark` in akka-bench-jmh